### PR TITLE
Adding UDP socket close functionality

### LINKED ISSUE #475: Udp rebinding / closing
It doesn't appear that smoltcp provides a mechanism to close or un-bind a UDP port (unless I'm missing something?).

As a result of this, once a UDP port has been bound, it may not be used for any other activities.

It would be beneficial to allow UDP ports to be re-bound to different sockets

### PR DESCRIPTION
**Pull request by ryan-summers**
This PR fixes #475 by adding a `close()` method to UDP sockets to unbind from the current endpoint.

### CODE REVIEW

File: src/socket/udp.rs [hunk 1]
```diff
@@ -146,6 +146,11 @@ impl<'a> UdpSocket<'a> {
         Ok(())
     }
 
+    /// Close the socket.
+    pub fn close(&mut self) {
+        self.endpoint = IpEndpoint::default();
```
Comment (Dirbaio) [inline src/socket/udp.rs:152]: This should wake rx_waker, tx_waker.

**Review by Dirbaio — CHANGES_REQUESTED** [review]
Sounds reasonable!

What happens with the packets in rx_buffer/tx_buffer when the socket is closed and reopened (possibly with a different endpoint)? Should they be cleared?

**Comment by ryan-summers** [comment]
I've added a new `PacketBuffer::clear()` API to allow us to flush the RX/TX buffer of the socket when closing it. Good catch there.

**Review by Dirbaio — APPROVED** [review]
Very nice! Thanks :)

**Review by Dirbaio — APPROVED** [review]

**Pull request merged**